Add Google calendar event handler

chime3 opened the pull request:
Closes #97

chime3 commented:
![calendar-event-tests](https://github.com/user-attachments/assets/ca36a41a-b8b2-40f8-ac88-1eaf1470cfad)
![calendar-tests](https://github.com/user-attachments/assets/b350b415-8a79-4610-b1ca-4b116ba62a59)

tahpot commented on src/providers/google/calendar-event.ts:
```diff
@@ -0,0 +1,280 @@
+      const insertedAt = new Date().toISOString();
+
+      const creator: Person = {
+        email: event.creator.email ?? 'info@example.com',
```
We can't have `info@example.com` values

chime3 replied:
> We can't have `info@example.com` values

Removed

tahpot reviewed (changes requested):
See comments

tahpot commented:
Test isn't passing for me.

![Screenshot 2024-09-30 at 6 36 01 PM](https://github.com/user-attachments/assets/0a29a334-7de4-4d4b-b229-f81d99a53ed0)

chime3 commented on src/providers/google/calendar-event.ts:
```diff
@@ -0,0 +1,280 @@
+
+      results.push({
+        _id: this.buildItemId(eventId),
+        name: event.summary ?? 'No event title',
```
> `No title`

Removed

chime3 commented on src/providers/google/calendar.ts:
```diff
@@ -0,0 +1,236 @@
+      
+      timeZone = CalendarHelpers.getUTCOffsetTimezone(timeZone);
+
+      const description = listItem.description ?? 'No description';
```
> Can these be blank, instead of `No xxx`?

Removed

chime3 commented on src/providers/google/calendar.ts:
```diff
@@ -0,0 +1,236 @@
+        break;
+      }
+
+      const summary = listItem.summary ?? 'No calendar title';
```
> `No title`

Removed

chime3 commented on src/providers/google/calendar-event.ts:
```diff
@@ -0,0 +1,280 @@
+        organizer,
+        location: event.location ?? 'No location',
+        description: event.description ?? 'No description',
+        status: event.status ?? 'Unkown',
```
> What are the options for status?
In general, the options are "confirmed", "tentative", "cancelled", but optional

chime3 commented on src/providers/google/calendar-event.ts:
```diff
@@ -0,0 +1,280 @@
+        creator,
+        organizer,
+        location: event.location ?? 'No location',
+        description: event.description ?? 'No description',
```
> Can we have `location`, `status` and `description` as empty values, instead of `No xxx`?

Fixed

chime3 commented on src/providers/google/calendar-event.ts:
```diff
@@ -0,0 +1,280 @@
+        sourceData: event,
+        sourceApplication: this.getProviderApplicationUrl(),
+        sourceId: eventId,
+        calendarId: "primary",
```
> This should be a reference to the `calendar._id` object associated with this event.

Fixed

tahpot commented on src/providers/google/interfaces.ts:
```diff
@@ -40,4 +40,27 @@ export enum YoutubeActivityType {
+
+export interface GoogleCalendarHandlerConfig extends BaseHandlerConfig {
+  calendarLimit?: number;  // Max number of calendar per sync
+  eventsPerCalendarLimit?: number; // Max number of event to process in a calendar
```
Rename toto `calendarBatchSize` and `eventBatchSize` for consistency with other handlers.

chime3 replied:
Fixed

tahpot commented on src/serverconfig.example.json:
```diff
@@ -86,16 +87,20 @@
             "clientId": "",
             "clientSecret": "",
-            "batchSize": 50,
+            "batchSize": 30,
```
Why was this changed to 30? I think it should be default of 50 for all handlers.

chime3 replied:
It was for test, and reverted

tahpot commented on src/serverconfig.example.json:
```diff
@@ -86,16 +87,20 @@
                 },
                 "google-drive-document": {
-                    "batchSize": 50
+                    "batchSize": 2
```
This should stay at 50

tahpot replied:
I've fixed this.

tahpot commented on src/providers/google/calendar-event.ts:
```diff
@@ -0,0 +1,390 @@
+          continue;
+        }
+
+        const summary = calendar.summary ?? "No calendar title";
```
Replace `No calendar title` with `No title`

chime3 replied:
Fixed

tahpot commented on src/providers/google/calendar-event.ts:
```diff
@@ -0,0 +1,390 @@
+      calendarId: calendar.sourceId,
+      maxResults: this.config.eventsPerCalendarLimit,
+      singleEvents: true,
+      orderBy: "startTime"
```
Google supports sorting by last updated, so we should use that so we automatically apply any changes to events.

chime3 replied:
Done

tahpot commented on src/providers/google/calendar-event.ts:
```diff
@@ -0,0 +1,390 @@
+      orderBy: "startTime"
+    };
+
+    if (range.startId && !isNaN(Date.parse(range.startId))) {
```
I would prefer if range.endId used pageToken from the Google API instead of a timestamp.

That is the exact placeholder of the next result set as specified by the Google API.

chime3 replied:
Done

tahpot commented on src/providers/google/calendar-event.ts:
```diff
@@ -0,0 +1,390 @@
+      query.timeMin = new Date(range.startId).toISOString();
+    }
+
+    if (range.endId && !isNaN(Date.parse(range.endId))) {
```
`range.endId` should be the ID of the record that is at the end of the range and if it is hit, then the loop ends. No need to limit in the Google API query.

chime3 replied:
Fixed

tahpot commented on src/providers/google/calendar-event.ts:
```diff
@@ -0,0 +1,390 @@
+  ) {
+    if (totalEvents === 0) {
+      syncPosition.status = SyncHandlerStatus.ENABLED;
+      syncPosition.syncMessage = "No new events found.";
```
`Stopping. No results found.` for consistency with Gmail.

chime3 replied:
Fixed

chime3 commented on src/providers/google/calendar-event.ts:
```diff
@@ -0,0 +1,390 @@
+        sourceApplication: this.getProviderApplicationUrl(),
+        sourceId: eventId,
+        schema: CONFIG.verida.schemas.EVENT,
+        calendarId: calendar.sourceId ?? "primary",
```
Removed

chime3 commented on src/providers/google/calendar-event.ts:
```diff
@@ -0,0 +1,390 @@
+
+    let nextPageToken: string | undefined;
+    let query: calendar_v3.Params$Resource$Calendarlist$List = {
+      maxResults: MAX_BATCH_SIZE,  // Fetch in batches up to the max limit
```
Fixed

chime3 commented:
### What's changed
- Replaced `timestamp` based item tracking with ID & `pageToken` tracking
- Modified common units for custom handler config
- Common unit tests cases are mainly checking basic item IDs 

### How to test
![image](https://github.com/user-attachments/assets/1fb3ef2f-28d9-4a2d-bcc7-adbefc7ade85)

tahpot reviewed (changes requested):
You can't make changes to `tests/common.tests.ts`. That is used by the other handler tests.

chime3 commented:
> You can't make changes to `tests/common.tests.ts`. That is used by the other handler tests.

Reverted and didn't use `runGenericTests` function of common unit tests, but used `buildTestObjects` and `checkItem` function.

Instead, added test cases in `calendar-event.tests.ts` file.
